src/cachinglayer-new.ts

```diff
@@ -0,0 +1,433 @@
+                  delete node.itemsMap[i];
+                }
+              }
+            }
+            return {statusCode: 200, body: node.body || node.itemsMap, contentType: node.contentType};
+          } else {
+            return {statusCode: 404};
+          }
+        });
+    }
+  }
+
+  put(path: string, body: any, contentType: ContentType): Promise<RSNodes> {
+    const paths = pathsFromRoot(path);
+
+    function _processNodes(nodePaths, nodes) {
+      try {
+        for (let i = 0, len = nodePaths.length; i < len; i++) {
+          const nodePath = nodePaths[i];
+          let node = nodes[nodePath];
+          let previous;
+
+          if(!node) {
+            nodes[nodePath] = node = makeNode(nodePath);
+          }
+
+          // Document
+          if(i === 0) {
+            previous = getLatest(node);
+            node.local = {
+              body: body,
+              contentType: contentType,
+              previousBody: (previous ? previous.body : undefined),
+              previousContentType: (previous ? previous.contentType : undefined),
+            };
+          }
+          // Folder
+          else {
+            const itemName = nodePaths[i - 1].substring(nodePath.length);
+            node = updateFolderNodeWithItemName(node, itemName);
+          }
+        }
+        return nodes;
+      } catch (e) {
+        log('[Cachinglayer] Error during PUT', nodes, e);
+        throw e;
+      }
+    }
+
+    return this._updateNodes(paths, _processNodes);
+  }
+
+  delete(path: string): unknown {
+    const paths = pathsFromRoot(path);
+
+    return this._updateNodes(paths, function (nodePaths, nodes) {
+      for (let i = 0, len = nodePaths.length; i < len; i++) {
+        const nodePath = nodePaths[i];
+        const node = nodes[nodePath];
+        let previous;
+
+        if(!node) {
+          console.error('Cannot delete non-existing node ' + nodePath);
+          continue;
+        }
+
+        if(i === 0) {
+          // Document
+          previous = getLatest(node);
+          node.local = {
+            body: false,
+            previousBody: (previous ? previous.body : undefined),
+            previousContentType: (previous ? previous.contentType : undefined),
+          };
+        } else {
+          // Folder
+          if(!node.local) {
+            node.local = deepClone(node.common);
+          }
+          const itemName = nodePaths[i - 1].substring(nodePath.length);
+          delete node.local.itemsMap[itemName];
+
+          if(Object.getOwnPropertyNames(node.local.itemsMap).length > 0) {
+            // This folder still contains other items, don't remove any further ancestors
+            break;
+          }
+        }
+      }
+      return nodes;
+    });
+  }
+
+  flush(path: string): unknown {
+
+    return this._getAllDescendentPaths(path).then((paths) => {
+      return this.getNodes(paths);
+    }).then(function (nodes: RSNodes) {
```

Comment:
> Also i think it would be nice to use `async/await` everywhere to make the code involving Promises cleanner and easier to understand

Totally agree. My fingers are yearning for the refactoring bonanza after the TS conversion. :)